subsys/power/device_pm.c

```diff
@@ -0,0 +1,301 @@
+	}
+
+	/* Notify the driver through signal */
+	if (dev->pm->signal) {
+		k_poll_signal_raise(dev->pm->signal, dev->pm->pm_status);
+	}
+
+	return ret;
+}
+
+/*
+ * Device PM worker to handle the Async PM requests
+ * from device_pm_get/device_pm_put API's. The worker
+ * will call the device_set_power_state() to put the
+ * device in desired power state and notify the caller
+ * through signal mechanism.
+ */
+static void pm_work_handler(struct k_work *work)
+{
+	struct device_pm *pm = CONTAINER_OF(work,
+					struct device_pm, dwork);
+	struct device *dev = pm->dev;
+	int ret;
+	u32_t request;
+
+	/* Set async status to in progress */
+	k_sem_take(&dev->pm->lock, K_FOREVER);
+	request = dev->pm->async_request;
+	dev->pm->async_req_status =  DEVPM_ASYNC_REQ_INPROG;
+	k_sem_give(&dev->pm->lock);
+
+	/* Certain devices may take lot of time to switch
+	 * between power states. So to avoid blocking of
+	 * Async API's for longer periods we are using a
+	 * dedicated lock to control power state transitins
+	 * between Sync and Async calls.
+	 */
+	k_sem_take(&dev->pm->pm_lock, K_FOREVER);
+	ret = set_dev_pm_state(dev, request);
+	if (ret) {
+		LOG_ERR("Set power state(%d) error(%d)\n", request, ret);
+	}
+	k_sem_give(&dev->pm->pm_lock);
+
+	/* Set async status to done */
+	k_sem_take(&dev->pm->lock, K_FOREVER);
+	dev->pm->async_req_status =  DEVPM_ASYNC_REQ_DONE;
+	k_sem_give(&dev->pm->lock);
+}
+
+/*
+ * Set the device PM state to target PM state based on
+ * the current state. If the PM request is of Async type
+ * then a work item will be submitted to PM worker queue
+ * else it will call device_set_power_state() to set the
+ * device PM state to target state.
+ */
+static int device_pm_state_handler(struct device *dev,
+				   u32_t target_state, u32_t pm_flags)
+{
+	u32_t current_state;
+	int ret = DEVICE_PM_RET_SUCCESS;
+
+	/* Get current device power state */
+	ret = device_get_power_state(dev, &current_state);
+	if (ret) {
+		LOG_ERR("Get power state error(%d)\n", ret);
+		goto handler_out;
+	}
+
+	/* Check if current power state is same as target
+	 * state so that we can skip processing new request.
+	 */
+	if (current_state == target_state) {
+		if (pm_flags & DEVICE_PM_ASYNC) {
+			if (dev->pm->async_request == target_state) {
+				goto handler_out;
+			}
+		} else {
+			goto handler_out;
+		}
+	}
+
+	/* Schedule the worker for new async request */
+	if (pm_flags & DEVICE_PM_ASYNC) {
+		k_delayed_work_cancel(&dev->pm->dwork);
+		dev->pm->async_request = target_state;
+		dev->pm->async_req_status = DEVPM_ASYNC_REQ_QUEUED;
+		k_delayed_work_submit_to_queue(&devpm_workq,
+					&dev->pm->dwork, dev->pm->timeout);
+
+		return DEVICE_PM_RET_QUEUED;
+	}
+
+	/* Certain devices may take lot of time to switch
+	 * between power states. So to avoid blocking of
+	 * Async API's for longer periods we are using a
+	 * dedicated lock to control power state transitins
+	 * between Sync and Async calls.
+	 */
+	k_sem_take(&dev->pm->pm_lock, K_FOREVER);
```

Comment:
Here you can wait for `pm_lock` (good), holding `lock` (bad).

Reply:
if we don't hold the lock then the async worker will also try to change the state.

Reply:
I think that we have to decouple requests from the entity which do the change. All requests will be asynchronous (sync API will just wait for proper event after requesting change). Unfortunately this will need dedicated thread (we already have one for workq) unless we change device_set_power_state API.

Reply:
yes, I think we will end up there :)

Reply:
I personally would recommend making device_set_power_state  asynchronous. Some devices might  not have memory for additional thread stack.

Reply:
How do I make it asynchronous? queue the work on the same pm thread? or have a different thread?

Reply:
Are you asking the device_set_power_state()? I just would add a callback which should be called when the device will change the state. If another device_set_power_state() will be called during transition, the call will return -EINPROGRESS. If device is busy, and cannot change state at the moment, it could return -EBUSY.

Reply:
Sorry, I didn't get,  are you suggesting to change the  device_set_power_state() implementation? do we need to add the callback in device_pm struct?

Reply:
I would start from changing the  device_set_power_state() - to avoid dedicated thread in device PM management.

Reply:
Ok, will work in that direction.

Reply:
> Ok, will work in that direction.
> 
> @pizi-nordic Even if we want to make the device_set_power_state() asynchronous we would still need to queue the work either on system thread or pm thread or a new thread. This can be done from device pm code itself right. why do we need to change the device_set_power_state() API?

@pizi-nordic I have addressed all your comments except this one. I will spend a bit more time on this and get back to you by tomorrow.

Reply:
If you make the device_set_power_state() asynchronous you will be able to avoid dedicated thread. All logic might be triggered by either device_get()/device_put() or device_set_power_state() callback.

Reply:
@pizi-nordic I have rewritten some of the device_pm logic to decouple async request queuing and device set power state. Please have a look.